client/mainmenu/CHighScoreScreen.cpp

```diff
@@ -0,0 +1,399 @@
+/*
+ * CHighScoreScreen.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "StdInc.h"
+
+#include "CHighScoreScreen.h"
+#include "../gui/CGuiHandler.h"
+#include "../gui/WindowHandler.h"
+#include "../gui/Shortcut.h"
+#include "../widgets/TextControls.h"
+#include "../widgets/Buttons.h"
+#include "../widgets/Images.h"
+#include "../widgets/MiscWidgets.h"
+#include "../windows/InfoWindows.h"
+#include "../render/Canvas.h"
+
+#include "../CGameInfo.h"
+#include "../CVideoHandler.h"
+#include "../CMusicHandler.h"
+#include "../../lib/CGeneralTextHandler.h"
+#include "../../lib/CConfigHandler.h"
+#include "../../lib/CCreatureHandler.h"
+#include "../../lib/constants/EntityIdentifiers.h"
+#include "../../lib/TextOperations.h"
+
+#include "vstd/DateUtils.h"
+
+auto HighScoreCalculation::calculate()
+{
+	struct result
+	{
+		int basic;
+		int total;
+		int sumDays;
+		bool cheater;
+	};
+	
+	std::vector<int> scoresBasic;
+	std::vector<int> scoresTotal;
+	double sumBasic = 0;
+	double sumTotal = 0;
+	int sumDays = 0;
+	bool cheater = false;
+	for(auto & param : parameters)
+	{
+		double tmp = 200 - (param.day + 10) / (param.townAmount + 5) + (param.allDefeated ? 25 : 0) + (param.hasGrail ? 25 : 0);
+		scoresBasic.push_back(static_cast<int>(tmp));
+		sumBasic += tmp;
+		if(param.difficulty == 0)
+			tmp *= 0.8;
+		if(param.difficulty == 1)
+			tmp *= 1.0;
+		if(param.difficulty == 2)
+			tmp *= 1.3;
+		if(param.difficulty == 3)
+			tmp *= 1.6;
+		if(param.difficulty == 4)
+			tmp *= 2.0;
+		scoresTotal.push_back(static_cast<int>(tmp));
+		sumTotal += tmp;
+		sumDays += param.day;
+		if(param.usedCheat)
+			cheater = true;
+	}
```

Comment:
```suggestion
	const std::array<double, 5> difficultyMultipliers{0.8, 1.0, 1.3, 1.6, 2.0}; 
	for(auto & param : parameters)
	{
		double tmp = 200 - (param.day + 10) / (param.townAmount + 5) + (param.allDefeated ? 25 : 0) + (param.hasGrail ? 25 : 0);
		scores.emplace_back(tmp, tmp * difficultyMultipliers.at(param.difficulty), param.day, param.usedCheat);
		summary.basic += scores.back().basic * 5.0 / parameters.size();
		summary.total+= scores.back().total * 5.0 / parameters.size();
		summary.sumDays += param.day;
		summary.cheater |= param.usedCheat;
	}
```

Reply:
even better: get rid of scores array and have only `Result firstResult`. Then in condition below
```c++
if(parameters.size() == 1)
    return firstResult;
return summary;
```